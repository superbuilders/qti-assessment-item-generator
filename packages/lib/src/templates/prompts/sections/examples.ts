import { existsSync, readFileSync } from "node:fs"
import * as path from "node:path"
import { resolveLibPath } from "@/internal/paths"

const POSITIVE_EXAMPLE_CACHE = new Map<string, string>()
const NEGATIVE_EXAMPLE_CACHE = new Map<string, string>()

const POSITIVE_ROOT = resolveLibPath("templates/prompts/examples/positive")
const POSITIVE_TEMPLATES_DIR = path.join(POSITIVE_ROOT, "templates")
const POSITIVE_NOTES_DIR = path.join(POSITIVE_ROOT, "notes")

const NEGATIVE_ROOT = resolveLibPath("templates/prompts/examples/negative")
const NEGATIVE_TEMPLATES_DIR = path.join(NEGATIVE_ROOT, "templates")
const NEGATIVE_NOTES_DIR = path.join(NEGATIVE_ROOT, "notes")

export function createPositiveExampleSection(names: readonly string[]): string {
	if (names.length === 0) return ""
	const rendered = names
		.map((name) => loadPositiveExample(name))
		.filter((block) => block.length > 0)
		.join("\n\n")
	if (rendered.length === 0) return ""
	return `### POSITIVE_EXAMPLES
<examples type="positive">
${rendered}
</examples>`
}

export function createNegativeExampleSection(names: readonly string[]): string {
	if (names.length === 0) return ""
	const rendered = names
		.map((name) => loadNegativeExample(name))
		.filter((block) => block.length > 0)
		.join("\n\n")
	if (rendered.length === 0) return ""
	return `### NEGATIVE_EXAMPLES
<examples type="negative">
${rendered}
</examples>`
}

function loadPositiveExample(name: string): string {
	const cached = POSITIVE_EXAMPLE_CACHE.get(name)
	if (cached) return cached

	const codePath = path.join(POSITIVE_TEMPLATES_DIR, `${name}.ts`)
	if (!existsSync(codePath)) {
		return ""
	}

	const notesPath = path.join(POSITIVE_NOTES_DIR, `${name}.md`)
	const code = readFileSync(codePath, "utf-8")
	const notes = existsSync(notesPath)
		? readFileSync(notesPath, "utf-8").trim()
		: ""

	const rendered = `<example kind="positive" name="${name}" source="${codePath}">
<code>
${code}
</code>
${notes ? `<notes>\n${notes}\n</notes>` : "<notes />"}
</example>`
	POSITIVE_EXAMPLE_CACHE.set(name, rendered)
	return rendered
}

function loadNegativeExample(name: string): string {
	const existing = NEGATIVE_EXAMPLE_CACHE.get(name)
	if (existing) return existing

	const codePath = path.join(NEGATIVE_TEMPLATES_DIR, `${name}.ts`)
	if (!existsSync(codePath)) {
		return ""
	}
	const notesPath = path.join(NEGATIVE_NOTES_DIR, `${name}.md`)
	const code = readFileSync(codePath, "utf-8")
	const notes = existsSync(notesPath)
		? readFileSync(notesPath, "utf-8").trim()
		: ""
	const rendered = `<example kind="negative" name="${name}" source="${codePath}">
<code>
${code}
</code>
${notes ? `<notes>\n${notes}\n</notes>` : "<notes />"}
</example>`
	NEGATIVE_EXAMPLE_CACHE.set(name, rendered)
	return rendered
}
