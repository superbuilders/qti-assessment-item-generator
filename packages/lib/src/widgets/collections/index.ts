import { allWidgetsCollection } from "@/widgets/collections/all"
import { fourthGradeMathCollection } from "@/widgets/collections/fourth-grade-math"
import { mathCoreCollection } from "@/widgets/collections/math-core"
import { scienceCollection } from "@/widgets/collections/science"
import { simpleVisualCollection } from "@/widgets/collections/simple-visual"
import { teksMath4Collection } from "@/widgets/collections/teks-math-4"

export const widgetCollections: {
	readonly all: typeof allWidgetsCollection
	readonly science: typeof scienceCollection
	readonly "simple-visual": typeof simpleVisualCollection
	readonly "fourth-grade-math": typeof fourthGradeMathCollection
	readonly "teks-math-4": typeof teksMath4Collection
	readonly "math-core": typeof mathCoreCollection
} = {
	all: allWidgetsCollection,
	science: scienceCollection,
	"simple-visual": simpleVisualCollection,
	"fourth-grade-math": fourthGradeMathCollection,
	"teks-math-4": teksMath4Collection,
	"math-core": mathCoreCollection
}

export type { WidgetCollection } from "@/widgets/collections/types"
