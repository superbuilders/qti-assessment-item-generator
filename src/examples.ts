import type { AssessmentItemInput } from "@/core/item"
import type { allWidgetsCollection } from "@/widgets/collections/all"
import type { WidgetTypeTupleFrom } from "@/widgets/collections/types"

export const allExamples: AssessmentItemInput<WidgetTypeTupleFrom<typeof allWidgetsCollection>>[] = []
