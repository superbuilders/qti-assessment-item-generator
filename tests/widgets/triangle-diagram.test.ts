import { describe, expect, test } from "bun:test"
import * as errors from "@superbuilders/errors"
import * as logger from "@superbuilders/slog"
import { triangleDiagramExamples } from "../../examples/triangle-diagram"
import type { WidgetInput } from "../../src/widgets/registry"
import { generateWidget } from "../../src/widgets/widget-generator"

describe("Widget: triangle-diagram", () => {
	const examples: WidgetInput[] = triangleDiagramExamples

	examples.forEach((props, index) => {
		test(`should produce consistent output for example #${index + 1}`, async () => {
			const result = await errors.try(generateWidget(props))
			if (result.error) {
				logger.error("widget generation failed", { error: result.error, index })
				throw result.error
			}
			expect(result.data).toMatchSnapshot()
		})
	})
})
