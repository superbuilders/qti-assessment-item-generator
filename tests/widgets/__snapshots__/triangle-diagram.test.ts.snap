// Bun Snapshot v1, https://bun.sh/docs/test/snapshots

exports[`Widget: triangle-diagram should produce consistent output for example #1 1`] = `"<svg width="502" height="329" viewBox="-20 -32 502 329" xmlns="http://www.w3.org/2000/svg" font-family="sans-serif" font-size="14"><path d="M 159.28509719458307 211 A 35.22 35.22 0 0 1 109.7398326254534 243.1750710181724" fill="none" stroke="#1fab54" stroke-width="2"/><text x="153.05078663808283" y="255.63404762605546" font-size="14" text-anchor="middle" dominant-baseline="middle" font-weight="700" fill="#000000">114°</text><path d="M 298.6549028054169 211 A 39.28 39.28 0 0 1 315.969805557086 178.43540415003798" fill="none" stroke="#11accd" stroke-width="2"/><text x="287.35966468645756" y="184.1086688836242" font-size="14" text-anchor="middle" dominant-baseline="middle" font-weight="700" fill="#000000">56°</text><path d="M 230.9905863303585 52.44853058980433 A 39.13999999999999 39.13999999999999 0 0 1 193.18410387852666 55.75616921213139" fill="none" stroke="#7854ab" stroke-width="2"/><text x="214.08385522911468" y="76.92256504896233" font-size="14" text-anchor="middle" dominant-baseline="middle" font-weight="700" fill="#000000">x</text><polygon points="124.06509719458307,211 337.9349028054169,211 209.10377608851348,20" fill="none" stroke="#000000" stroke-width="2"/><line x1="98.55349352640395" y1="268.3" x2="124.06509719458307" y2="211" stroke="#000000" stroke-width="2"/><line x1="124.06509719458307" y1="211" x2="209.10377608851348" y2="20" stroke="#000000" stroke-width="2"/><circle cx="124.06509719458307" cy="211" r="5" fill="#000000"/><circle cx="337.9349028054169" cy="211" r="5" fill="#000000"/><circle cx="209.10377608851348" cy="20" r="5" fill="#000000"/><text x="110.99376035422243" y="190.8719063693098" font-size="16" text-anchor="middle" dominant-baseline="middle" font-weight="700" fill="#000000">A</text><text x="359.12564503403115" y="222.2673175068614" font-size="16" text-anchor="middle" dominant-baseline="middle" font-weight="700" fill="#000000">B</text><text x="207.0120382625697" y="-3.908672754201895" font-size="16" text-anchor="middle" dominant-baseline="middle" font-weight="700" fill="#000000">C</text><circle cx="98.55349352640395" cy="268.3" r="5" fill="#000000"/><text x="122.55349352640395" y="268.3" font-size="16" text-anchor="middle" dominant-baseline="middle" font-weight="700" fill="#000000">D</text></svg>"`;

exports[`Widget: triangle-diagram should produce consistent output for example #2 1`] = `"<svg width="502" height="328" viewBox="-20 -32 502 328" xmlns="http://www.w3.org/2000/svg" font-family="sans-serif" font-size="14"><path d="M 324.78901534588226 211 A 38.370000000000005 38.370000000000005 0 0 1 349.4084371218291 175.17851903530237" fill="none" stroke="#7854ab" stroke-width="2"/><text x="316.7030220932592" y="179.0716804245472" font-size="14" text-anchor="middle" dominant-baseline="middle" font-weight="700" fill="#000000">69°</text><path d="M 303.6668201475755 56.017532854262036 A 38.58 38.58 0 0 1 262.5608050359408 47.280179618177" fill="none" stroke="#11accd" stroke-width="2"/><text x="278.077341187649" y="75.34359124951872" font-size="14" text-anchor="middle" dominant-baseline="middle" font-weight="700" fill="#000000">66°</text><path d="M 132.59098465411776 211 A 33.75 33.75 0 0 1 74.97613078907179 234.86485386504597" fill="none" stroke="#1fab54" stroke-width="2"/><text x="118.64485227901116" y="258.8107658074591" font-size="14" text-anchor="middle" dominant-baseline="middle" font-weight="700" fill="#000000">x</text><polygon points="98.84098465411776,211 363.15901534588227,211 289.8409846541178,20" fill="none" stroke="#000000" stroke-width="2"/><line x1="41.54098465411776" y1="268.3" x2="98.84098465411776" y2="211" stroke="#000000" stroke-width="2"/><line x1="98.84098465411776" y1="211" x2="289.8409846541178" y2="20" stroke="#000000" stroke-width="2"/><circle cx="98.84098465411776" cy="211" r="5" fill="#000000"/><circle cx="363.15901534588227" cy="211" r="5" fill="#000000"/><circle cx="289.8409846541178" cy="20" r="5" fill="#000000"/><text x="89.6565822773556" y="188.82689121972913" font-size="16" text-anchor="middle" dominant-baseline="middle" font-weight="700" fill="#000000">A</text><text x="382.9380438728106" y="224.593749686196" font-size="16" text-anchor="middle" dominant-baseline="middle" font-weight="700" fill="#000000">B</text><text x="294.830865233744" y="-3.47554241761134" font-size="16" text-anchor="middle" dominant-baseline="middle" font-weight="700" fill="#000000">C</text><circle cx="41.54098465411776" cy="268.3" r="5" fill="#000000"/><text x="65.54098465411775" y="268.3" font-size="16" text-anchor="middle" dominant-baseline="middle" font-weight="700" fill="#000000">D</text></svg>"`;

exports[`Widget: triangle-diagram should produce consistent output for example #3 1`] = `"<svg width="502" height="355" viewBox="-20 -59 502 355" xmlns="http://www.w3.org/2000/svg" font-family="sans-serif" font-size="14"><path d="M 303.4482066865706 59.32411755917558 A 41.589999999999996 41.589999999999996 0 0 1 289.1594443119788 50.90739329160482" fill="none" stroke="#11accd" stroke-width="2"/><text x="347.2327972792721" y="-31.344481670710515" font-size="14" text-anchor="middle" dominant-baseline="middle" font-weight="700" fill="#000000">23°</text><path d="M 171.6695771268932 181.39311015298063 A 39.84 39.84 0 0 1 184.85141376955627 211" fill="none" stroke="#11accd" stroke-width="2"/><text x="197.8508830396043" y="187.47435256449572" font-size="14" text-anchor="middle" dominant-baseline="middle" font-weight="700" fill="#000000">48°</text><path d="M 178.97141376955625 211 A 33.959999999999994 33.959999999999994 0 0 1 122.28773837760946 236.2371982732123" fill="none" stroke="#1fab54" stroke-width="2"/><text x="166.14544974377486" y="258.46782197910954" font-size="14" text-anchor="middle" dominant-baseline="middle" font-weight="700" fill="#000000">x</text><polygon points="145.01141376955627,211 251.22201209211767,211 316.98858623044373,20" fill="none" stroke="#000000" stroke-width="2"/><line x1="93.41826203129003" y1="268.3" x2="145.01141376955627" y2="211" stroke="#000000" stroke-width="2"/><line x1="145.01141376955627" y1="211" x2="316.98858623044373" y2="20" stroke="#000000" stroke-width="2"/><circle cx="145.01141376955627" cy="211" r="5" fill="#000000"/><circle cx="251.22201209211767" cy="211" r="5" fill="#000000"/><circle cx="316.98858623044373" cy="20" r="5" fill="#000000"/><text x="135.24973433573706" y="189.07490901657758" font-size="16" text-anchor="middle" dominant-baseline="middle" font-weight="700" fill="#000000">A</text><text x="275.22201209211767" y="211" font-size="16" text-anchor="middle" dominant-baseline="middle" font-weight="700" fill="#000000">B</text><text x="329.16950694150063" y="-0.6790998505966108" font-size="16" text-anchor="middle" dominant-baseline="middle" font-weight="700" fill="#000000">C</text><circle cx="93.41826203129003" cy="268.3" r="5" fill="#000000"/><text x="117.41826203129003" y="268.3" font-size="16" text-anchor="middle" dominant-baseline="middle" font-weight="700" fill="#000000">D</text></svg>"`;

exports[`Widget: triangle-diagram should produce consistent output for example #4 1`] = `"<svg width="502" height="317" viewBox="-20 -59 502 317" xmlns="http://www.w3.org/2000/svg" font-family="sans-serif" font-size="14"><path d="M 171.6695771268932 181.39311015298063 A 39.84 39.84 0 0 1 184.85141376955627 211" fill="none" stroke="#11accd" stroke-width="2"/><text x="197.8508830396043" y="187.47435256449572" font-size="14" text-anchor="middle" dominant-baseline="middle" font-weight="700" fill="#000000">48°</text><path d="M 303.4482066865706 59.32411755917558 A 41.589999999999996 41.589999999999996 0 0 1 289.1594443119788 50.90739329160482" fill="none" stroke="#1fab54" stroke-width="2"/><text x="347.2327972792721" y="-31.344481670710515" font-size="14" text-anchor="middle" dominant-baseline="middle" font-weight="700" fill="#000000">23°</text><path d="M 215.65201209211767 211 A 35.57 35.57 0 0 1 262.8024713461587 177.36790426593228" fill="none" stroke="#7854ab" stroke-width="2"/><text x="220.1137547546826" y="167.38783168165202" font-size="14" text-anchor="middle" dominant-baseline="middle" font-weight="700" fill="#000000">x</text><polygon points="145.01141376955627,211 251.22201209211767,211 316.98858623044373,20" fill="none" stroke="#000000" stroke-width="2"/><circle cx="145.01141376955627" cy="211" r="5" fill="#000000"/><circle cx="251.22201209211767" cy="211" r="5" fill="#000000"/><circle cx="316.98858623044373" cy="20" r="5" fill="#000000"/><text x="123.08632278613385" y="220.7616794338192" font-size="16" text-anchor="middle" dominant-baseline="middle" font-weight="700" fill="#000000">A</text><text x="265.15888302918023" y="230.53877244055167" font-size="16" text-anchor="middle" dominant-baseline="middle" font-weight="700" fill="#000000">B</text><text x="329.16950694150063" y="-0.6790998505966108" font-size="16" text-anchor="middle" dominant-baseline="middle" font-weight="700" fill="#000000">C</text></svg>"`;

exports[`Widget: triangle-diagram should produce consistent output for example #5 1`] = `"<svg width="558" height="271" viewBox="-72 -20 558 271" xmlns="http://www.w3.org/2000/svg" font-family="sans-serif" font-size="14"><path d="M 58.958311197728236 181.54061685794736 A 41.730000000000004 41.730000000000004 0 0 1 61.730000000000004 196.49531139247273" fill="none" stroke="#11accd" stroke-width="2"/><text x="-38.729815628795016" y="207.38023933011868" font-size="14" text-anchor="middle" dominant-baseline="middle" font-weight="700" fill="#000000">21°</text><path d="M 294.1229448066958 196.49531139247273 A 34.45 34.45 0 0 1 348.3326530389893 168.27552346671695" fill="none" stroke="#1fab54" stroke-width="2"/><text x="304.3542300425182" y="149.971593192275" font-size="14" text-anchor="middle" dominant-baseline="middle" font-weight="700" fill="#000000">125°</text><path d="M 418.58660986815033 67.94247505540393 A 40.81999999999999 40.81999999999999 0 0 1 403.8912469903842 49.133268307966425" fill="none" stroke="#7854ab" stroke-width="2"/><text x="395.64920747285265" y="70.71789658618252" font-size="14" text-anchor="middle" dominant-baseline="middle" font-weight="700" fill="#000000">x</text><polygon points="20,196.49531139247273 328.5729448066958,196.49531139247273 442,34.50468860752727" fill="none" stroke="#000000" stroke-width="2"/><circle cx="20" cy="196.49531139247273" r="5" fill="#000000"/><circle cx="328.5729448066958" cy="196.49531139247273" r="5" fill="#000000"/><circle cx="442" cy="34.50468860752727" r="5" fill="#000000"/><text x="-3.59811778153491" y="200.86896400428427" font-size="16" text-anchor="middle" dominant-baseline="middle" font-weight="700" fill="#000000">A</text><text x="339.65491152433657" y="217.78357138875003" font-size="16" text-anchor="middle" dominant-baseline="middle" font-weight="700" fill="#000000">B</text><text x="460.91225808656134" y="19.72881319971146" font-size="16" text-anchor="middle" dominant-baseline="middle" font-weight="700" fill="#000000">C</text></svg>"`;

exports[`Widget: triangle-diagram should produce consistent output for example #6 1`] = `"<svg width="502" height="287" viewBox="-20 -30 502 287" xmlns="http://www.w3.org/2000/svg" font-family="sans-serif" font-size="14"><path d="M 317.12278046474916 211 A 40.61 40.61 0 0 1 325.3001924017286 186.56029190979532" fill="none" stroke="#11accd" stroke-width="2"/><text x="302.1515310331138" y="192.40277408809757" font-size="14" text-anchor="middle" dominant-baseline="middle" font-weight="700" fill="#000000">37°</text><path d="M 135.80533582701844 43.76567526427439 A 39.49 39.49 0 0 1 104.26721953525086 59.49" fill="none" stroke="#1fab54" stroke-width="2"/><text x="129.91913181093378" y="71.44977644850042" font-size="14" text-anchor="middle" dominant-baseline="middle" font-weight="700" fill="#000000">53°</text><polygon points="104.26721953525085,211 357.73278046474917,211 104.26721953525086,20" fill="none" stroke="#000000" stroke-width="2"/><line x1="104.26721953525085" y1="211" x2="196.06771149736028" y2="89.17663251947657" stroke="#777777" stroke-width="1.5" stroke-dasharray="4 3"/><line x1="184.88681435669818" y1="80.75122219534789" x2="176.4614040325695" y2="91.93211933600999" stroke="#000000" stroke-width="2"/><line x1="187.6423011732316" y1="100.35752966013867" x2="176.4614040325695" y2="91.93211933600999" stroke="#000000" stroke-width="2"/><text x="161.34836265696765" y="158.51372658386697" font-size="12" text-anchor="middle" dominant-baseline="middle" fill="#000000">8</text><circle cx="104.26721953525085" cy="211" r="5" fill="#000000"/><circle cx="357.73278046474917" cy="211" r="5" fill="#000000"/><circle cx="104.26721953525086" cy="20" r="5" fill="#000000"/><text x="121.237782283728" y="227.97056274847714" font-size="16" text-anchor="middle" dominant-baseline="middle" font-weight="700" fill="#000000">A</text><text x="380.49254818969797" y="218.6153117537222" font-size="16" text-anchor="middle" dominant-baseline="middle" font-weight="700" fill="#000000">B</text><text x="93.55847202061545" y="-1.4784246784486044" font-size="16" text-anchor="middle" dominant-baseline="middle" font-weight="700" fill="#000000">C</text><text x="231" y="231" font-size="12" text-anchor="middle" dominant-baseline="middle" fill="#000000">3</text><text x="243.03630046304096" y="99.52728979905415" font-size="12" text-anchor="middle" dominant-baseline="middle" fill="#000000">5</text><text x="84.26721953525086" y="115.5" font-size="12" text-anchor="middle" dominant-baseline="middle" fill="#000000">x</text></svg>"`;

exports[`Widget: triangle-diagram should produce consistent output for example #7 1`] = `"<svg width="502" height="287" viewBox="-20 -29 502 287" xmlns="http://www.w3.org/2000/svg" font-family="sans-serif" font-size="14"><path d="M 144.51841861106618 185.5314037742131 A 40.47 40.47 0 0 1 153.53732155090256 211" fill="none" stroke="#11accd" stroke-width="2"/><text x="168.18356953506225" y="191.48231294060142" font-size="14" text-anchor="middle" dominant-baseline="middle" font-weight="700" fill="#000000">39°</text><path d="M 340.51849232170275 59.585633401697116 A 40.47 40.47 0 0 1 317.4815813889339 45.46859622578692" fill="none" stroke="#1fab54" stroke-width="2"/><text x="318.3821873701559" y="69.85387040978378" font-size="14" text-anchor="middle" dominant-baseline="middle" font-weight="700" fill="#000000">39°</text><path d="M 272.2743751701233 211 A 36.06 36.06 0 0 1 315.83167074101164 175.72799751753897" fill="none" stroke="#7854ab" stroke-width="2"/><text x="274.31331482996904" y="168.98748932363617" font-size="14" text-anchor="middle" dominant-baseline="middle" font-weight="700" fill="#000000">x</text><polygon points="113.06732155090256,211 308.3343751701233,211 348.93267844909747,20" fill="none" stroke="#000000" stroke-width="2"/><line x1="308.3343751701233" y1="211" x2="231.00000000000006" y2="115.49999999999996" stroke="#777777" stroke-width="1.5" stroke-dasharray="4 3"/><line x1="220.11995653960247" y1="124.31048547469769" x2="228.93044201430018" y2="135.19052893509527" stroke="#000000" stroke-width="2"/><line x1="239.81048547469777" y1="126.38004346039754" x2="228.93044201430018" y2="135.19052893509527" stroke="#000000" stroke-width="2"/><text x="280.5472310454593" y="154.43951452530226" font-size="12" text-anchor="middle" dominant-baseline="middle" fill="#000000">21</text><circle cx="113.06732155090256" cy="211" r="5" fill="#000000"/><circle cx="308.3343751701233" cy="211" r="5" fill="#000000"/><circle cx="348.93267844909747" cy="20" r="5" fill="#000000"/><text x="90.44392576469028" y="219.0113646216105" font-size="16" text-anchor="middle" dominant-baseline="middle" font-weight="700" fill="#000000">A</text><text x="323.4380645553194" y="229.6515030749673" font-size="16" text-anchor="middle" dominant-baseline="middle" font-weight="700" fill="#000000">B</text><text x="361.4726440022802" y="-0.4633639444982194" font-size="16" text-anchor="middle" dominant-baseline="middle" font-weight="700" fill="#000000">C</text><text x="210.7008483605129" y="231" font-size="12" text-anchor="middle" dominant-baseline="middle" fill="#000000">3.3</text><text x="348.1964788242865" y="119.65823381635518" font-size="12" text-anchor="middle" dominant-baseline="middle" fill="#000000">5.1</text><text x="218.41359217900325" y="99.95708077086059" font-size="12" text-anchor="middle" dominant-baseline="middle" fill="#000000">3.3</text></svg>"`;

exports[`Widget: triangle-diagram should produce consistent output for example #8 1`] = `"<svg width="518" height="271" viewBox="-28 -20 518 271" xmlns="http://www.w3.org/2000/svg" font-family="sans-serif" font-size="14"><path d="M 53.380445804776414 165.99865549981925 A 40.75 40.75 0 0 1 60.75 189.37189528112438" fill="none" stroke="#11accd" stroke-width="2"/><text x="76.03087085645834" y="171.70542956024832" font-size="14" text-anchor="middle" dominant-baseline="middle" font-weight="700" fill="#000000">35°</text><path d="M 401.25 189.37189528112438 A 40.75 40.75 0 0 1 408.61955419522354 165.99865549981928" fill="none" stroke="#1fab54" stroke-width="2"/><text x="385.96912914354164" y="171.70542956024838" font-size="14" text-anchor="middle" dominant-baseline="middle" font-weight="700" fill="#000000">35°</text><path d="M 260.0798975722592 61.99006820933776 A 35.5 35.5 0 0 1 201.9201024277408 61.99006820933775" fill="none" stroke="#7854ab" stroke-width="2"/><text x="231" y="95.12810471887562" font-size="14" text-anchor="middle" dominant-baseline="middle" font-weight="700" fill="#000000">x</text><polygon points="20,189.37189528112438 442,189.37189528112438 231,41.62810471887562" fill="none" stroke="#000000" stroke-width="2"/><line x1="231" y1="41.62810471887562" x2="231" y2="189.37189528112438" stroke="#777777" stroke-width="1.5" stroke-dasharray="4 3"/><line x1="245" y1="189.37189528112438" x2="245" y2="175.37189528112438" stroke="#000000" stroke-width="2"/><line x1="231" y1="175.37189528112438" x2="245" y2="175.37189528112438" stroke="#000000" stroke-width="2"/><circle cx="20" cy="189.37189528112438" r="5" fill="#000000"/><circle cx="442" cy="189.37189528112438" r="5" fill="#000000"/><circle cx="231" cy="41.62810471887562" r="5" fill="#000000"/><text x="-2.8892068179574437" y="196.58883446922695" font-size="16" text-anchor="middle" dominant-baseline="middle" font-weight="700" fill="#000000">A</text><text x="464.8892068179575" y="196.58883446922692" font-size="16" text-anchor="middle" dominant-baseline="middle" font-weight="700" fill="#000000">B</text><text x="231" y="17.62810471887562" font-size="16" text-anchor="middle" dominant-baseline="middle" font-weight="700" fill="#000000">C</text><text x="231" y="209.37189528112438" font-size="12" text-anchor="middle" dominant-baseline="middle" fill="#000000">8</text><foreignObject x="287.97152872702094" y="83.11695911422017" width="120" height="32"><!DOCTYPE html><div xmlns="http://www.w3.org/1999/xhtml" style="display:flex;align-items:flex-end;justify-content:center;width:100%;height:100%;line-height:1;font-family:sans-serif;color:#000000;"><math xmlns="http://www.w3.org/1998/Math/MathML" display="inline" style="font-size:17.6px;"><msqrt><mn>20</mn></msqrt></math></div></foreignObject><foreignObject x="54.028471272979075" y="83.11695911422017" width="120" height="32"><!DOCTYPE html><div xmlns="http://www.w3.org/1999/xhtml" style="display:flex;align-items:flex-end;justify-content:center;width:100%;height:100%;line-height:1;font-family:sans-serif;color:#000000;"><math xmlns="http://www.w3.org/1998/Math/MathML" display="inline" style="font-size:17.6px;"><msqrt><mn>20</mn></msqrt></math></div></foreignObject></svg>"`;

exports[`Widget: triangle-diagram should produce consistent output for example #9 1`] = `"<svg width="517" height="271" viewBox="-28 -20 517 271" xmlns="http://www.w3.org/2000/svg" font-family="sans-serif" font-size="14"><path d="M 50.94819550200671 178.05639165786684 A 40.4 40.4 0 0 1 60.4 204.02501108920302" fill="none" stroke="#11accd" stroke-width="2"/><text x="74.87804905389704" y="184.05103471898397" font-size="14" text-anchor="middle" dominant-baseline="middle" font-weight="700" fill="#000000">40°</text><path d="M 401.6 204.02501108920302 A 40.4 40.4 0 0 1 411.05180449799326 178.05639165786687" fill="none" stroke="#1fab54" stroke-width="2"/><text x="387.12195094610297" y="184.051034718984" font-size="14" text-anchor="middle" dominant-baseline="middle" font-weight="700" fill="#000000">40°</text><polygon points="20,204.02501108920302 442,204.02501108920302 230.99999999999997,26.974988910796952" fill="none" stroke="#000000" stroke-width="2"/><line x1="230.99999999999997" y1="26.974988910796952" x2="230.99999999999997" y2="204.02501108920302" stroke="#777777" stroke-width="1.5" stroke-dasharray="4 3"/><line x1="244.99999999999997" y1="204.02501108920302" x2="244.99999999999997" y2="190.02501108920302" stroke="#000000" stroke-width="2"/><line x1="230.99999999999997" y1="190.02501108920302" x2="244.99999999999997" y2="190.02501108920302" stroke="#000000" stroke-width="2"/><text x="216.99999999999997" y="115.49999999999999" font-size="12" text-anchor="middle" dominant-baseline="middle" fill="#000000">3</text><circle cx="20" cy="204.02501108920302" r="5" fill="#000000"/><circle cx="442" cy="204.02501108920302" r="5" fill="#000000"/><circle cx="230.99999999999997" cy="26.974988910796952" r="5" fill="#000000"/><text x="-2.5526228988617987" y="212.23349452901908" font-size="16" text-anchor="middle" dominant-baseline="middle" font-weight="700" fill="#000000">A</text><text x="464.5526228988618" y="212.23349452901905" font-size="16" text-anchor="middle" dominant-baseline="middle" font-weight="700" fill="#000000">B</text><text x="254.99999999999997" y="26.974988910796952" font-size="16" text-anchor="middle" dominant-baseline="middle" font-weight="700" fill="#000000">C</text><text x="231" y="224.02501108920302" font-size="12" text-anchor="middle" dominant-baseline="middle" fill="#000000">x</text><foreignObject x="289.3557521937308" y="84.17911113762042" width="120" height="32"><!DOCTYPE html><div xmlns="http://www.w3.org/1999/xhtml" style="display:flex;align-items:flex-end;justify-content:center;width:100%;height:100%;line-height:1;font-family:sans-serif;color:#000000;"><math xmlns="http://www.w3.org/1998/Math/MathML" display="inline" style="font-size:17.6px;"><msqrt><mn>45</mn></msqrt></math></div></foreignObject><foreignObject x="52.64424780626919" y="84.17911113762042" width="120" height="32"><!DOCTYPE html><div xmlns="http://www.w3.org/1999/xhtml" style="display:flex;align-items:flex-end;justify-content:center;width:100%;height:100%;line-height:1;font-family:sans-serif;color:#000000;"><math xmlns="http://www.w3.org/1998/Math/MathML" display="inline" style="font-size:17.6px;"><msqrt><mn>45</mn></msqrt></math></div></foreignObject></svg>"`;
